### PR TITLE
Create new TestPlan view with Patternfly

### PR DESCRIPTION
**Pull request by asankov**

### CODE REVIEW

File: tcms/templates/base.html [hunk 1]
```diff
@@ -9,6 +9,7 @@
     <title>Kiwi TCMS - {% block title %}{% trans "the leading open source test case management system" %}{% endblock %}</title>
     {% include 'patternfly.html' %}
     {% block head %}{% endblock %}
+    {% block extra_head %}{% endblock %}
```
Comment (atodorov) [inline tcms/templates/base.html]: Do you see the block above ?

File: tcms/testplans/templates/testplans/mutable.html [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                    </select>
+                </div>
+            </div>
+
+            <div class="form-group">
+                <label class="col-md-1 col-lg-1" for="id_parent_id">{% trans "Parent ID" %}</label>
+                <div class="col-md-3 col-lg-3">
+                    <input type="text" id="id_parent_id" name="parent_id" value="" class="form-control">
+                </div>
+                <label class="col-md-1 col-lg-1" for="id_product">{% trans "Reference link" %}</label>
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: id mismatch

File: tcms/testplans/templates/testplans/mutable.html [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                <div class="col-md-7 col-lg-7">
+                    <input type="text" id="id_summary" name="summary" value="" class="form-control">
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: id and name mismatch ? Doesn't this produce an error when you try to save the TestPlan ?

File: tcms/testplans/templates/testplans/mutable.html [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                </div>
+            </div>
+
+            <div class="form-group">
+                <label class="col-md-12 col-lg-12" for="messageArea" for="id_summary"> {% trans "Test plan document:" %}</label>
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: double for attribute and bogus id

File: tcms/testplans/templates/testplans/mutable.html [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    </div>
+
+<script src="{% static "grappelli/jquery/jquery.min.js" %}" type="text/javascript"></script>
+<script src="{% static "grappelli/js/grappelli.min.js" %}" type="text/javascript"></script>
+<script src="{% static "admin/js/admin/RelatedObjectLookups.js" %}" type="text/javascript"></script>
+
+<script src="{% static 'bootstrap-select/dist/js/bootstrap-select.min.js' %}" type="text/javascript"></script>
+    <script src="{% static 'js/jsonrpc.js' %}"></script>
+
+{#<script src="{% static 'testplans/js/mutable.js' %}" type="text/javascript"></script>#}
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: Pretty sure none of these js files are used ATM so whatever we need must be added in a separate commit.

File: tcms/testplans/templates/testplans/mutable.html (hunk 3 above)
Comment (asankov) [reply tcms/testplans/templates/testplans/mutable.html]: No. The `id` is for js/css related stuff, while `name` is for the form. It seems to be fine

Comment (atodorov) [reply tcms/testplans/templates/testplans/mutable.html]: I see you've updated it with the correct values :)

File: tcms/testplans/static/testplans/js/mutable.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+$(document).ready(() => {
```
[the remark this answers is not in the document]
Comment (asankov) [reply tcms/testplans/static/testplans/js/mutable.js]: If we indeed what to use that style and know that eventually eslint will enforce it I suggest we start doing it now. These are new js files, which should replace a big part of the old ones. From this point of view, a lot of the other js code will be deleted. So if we start to use eslint at that point we will have to refactor only the old chunks of code.

File: tcms/testplans/static/testplans/js/mutable.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    loadProductVersions();
+});
+
+$('#id_product').change(() => loadProductVersions());
+
+const loadProductVersions = () => {
```
[the remark this answers is not in the document]
Comment (asankov) [reply tcms/testplans/static/testplans/js/mutable.js]: should I just copy and paste it. I strived to write a brand new js file, without copying from the other 1500 lines long monstrosity. I think as this evolves we will write more functions and then when we encounter duplication will refactor, rather than copy the old logic, which is very generic, but hardly readable

File: tcms/testplans/static/testplans/js/mutable.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (atodorov) [reply tcms/testplans/static/testplans/js/mutable.js]: I tried eslint on this version (default config settings) and it reported an error.  Additionally eslint allows you to choose from 3 style guides by default and between 6 ECMAScript syntax versions. ATM we have no idea about how compatible are different browsers with the various ECMA script syntax versions nor what kind of browser/version combinations are in favor for our user base.

Classic style and being careful about compatibility, also consistency with the rest of the code is more important.

Feel free to open an issue to investigate the integration of eslint and possible refactoring of the rest of the JavaScript code and we can deal with that when we have the time and at least some information about browser versions used by others.

File: tcms/testplans/static/testplans/js/mutable.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (atodorov) [reply tcms/testplans/static/testplans/js/mutable.js]: > should I just copy and paste it.

hell no. Include the other .js file in the HTML before the current one and the function will become available for your onchange handler.

[the remark this answers is not in the document]
Comment (asankov) [reply tcms/testplans/static/testplans/js/mutable.js]: Sure, I could do that. But wasn't the purpose here to extract some of the old js code. This way we will see for possible duplicated and/or unused code. Just importing the old one would couple us to it and make the situation worse for the future

[the remark this answers is not in the document]
Comment (atodorov) [reply tcms/testplans/static/testplans/js/mutable.js]: This is exactly what we are doing. If you do `git log` & `git blame` on `tcms/static/js/utils.js` you will see that it is a new file and `update_version_select_from_product()`  is a callback intended to be used with the JSON RPC helper.

But we don't want to copy & paste the same function into another file which is one of the original problems that we are solving but instead we want to re-use the existing functions when these functions are good.

File: tcms/testplans/templates/testplans/mutable.html [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+{% block contents %}
+
+    <div class="container-fluid container-cards-pf">
+        <form class="form-horizontal" action="{% url 'plans-new' %}" enctype="multipart/form-data" method="post">
+            {% csrf_token %}
+
+            <div class="form-group">
+                <label class="col-md-1 col-lg-1" for="id_name">{% trans "Name" %}</label>
+                <div class="col-md-11 col-lg-11">
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: You are missing markup for error situations. This is missing for the following fields in the form:
name, product, product_version, type, parent, text, extra_link.  

For example in `testruns/mutable.html` we have the following pattern:
```
            <label class="col-md-1 col-lg-1" for="id_manager">{% trans "Manager" %}</label>
            <div class="col-md-3 col-lg-3 {% if form.manager.errors %}has-error{% endif %}">
                <input id="id_manager" name="manager" value="{{ form.manager.value }}" type="text"
                    class="form-control" placeholder="{% trans 'Username or email' %}" required>
                {{ form.manager.errors }}
            </div>
```

- add `has-error` class to the div tag defining the element
- under the input tag print the actual errors coming from the form

File: tcms/testplans/static/testplans/js/mutable.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+$(document).ready((function() {
+    update_version_select_from_product();
+
+    $('#id_product').change(update_version_select_from_product);
```
Comment (atodorov) [inline tcms/testplans/static/testplans/js/mutable.js]: Change to:

```
    $('#id_product').change(function() {
        $('#id_version').find('option').remove();
        update_version_select_from_product();
    });
```

to resolve the problem with the first element not being removed.

File: tcms/testplans/templates/testplans/mutable.html [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+                    <input type="text" id="id_name" name="name" value="{% firstof form.name.value "" %}" class="form-control">
+                </div>
+            </div>
+
+            <div class="form-group">
+                <div class="col-md-1 col-lg-1">
+                    <label for="id_product">{% trans "Product" %}</label>
+                    <a href="{% url 'admin:management_product_add' %}?_popup" id="add_id_product" alt="{% trans 'add new Product' %}" title="{% trans 'add new Product' %}">+</a>
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: For the '+' sign to work you need the following (again hints from `testruns/mutable.html`):

For the on-load handler you have to add:
```
    $('#add_id_product').click(function() {
        return showRelatedObjectPopup(this);
    });

    $('#add_id_version').click(function() {
        return showRelatedObjectPopup(this);
    });
```

This requires the following resources inside the template (before files coming from our project):

```
<script src="{% static "grappelli/jquery/jquery.min.js" %}" type="text/javascript"></script>
<script src="{% static "grappelli/js/grappelli.min.js" %}" type="text/javascript"></script>
<script src="{% static "admin/js/admin/RelatedObjectLookups.js" %}" type="text/javascript"></script>
```

This will make the popup for adding new product work.

File: tcms/testplans/templates/testplans/mutable.html [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+                        {% for product in form.product.field.queryset %}
+                            <option value="{{ product.pk }} {% if product.pk == form.product.value %}selected{% endif %}">
+                                {{ product.name }}
+                            </option>
+                        {% endfor %}
+                    </select>
+                </div>
+                <div class="col-md-1 col-lg-1">
+                    <label for="id_version">{% trans "Version" %}</label>
+                    <a href="{% url 'admin:management_version_add' %}?_popup"
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: This link will open the popup form to add new Version without the Product field being pre-selected. So it is possible for the user to create a Product version against a different product, not the one currently being selected.

So your on-product-change handler needs to append to this URL `&product=X` where `X` is the ID of the actual product selected from the Product drop down widget. In this way when the user clicks on the + button the popup that opens will have the same product pre-selected.

File: tcms/testplans/templates/testplans/mutable.html (hunk 8 above)
Comment (asankov) [reply tcms/testplans/templates/testplans/mutable.html]: Written this way it looks a bit ugly -
screenshot 2018-12-01 at 18 46 04

Putting `{{ form.name.errors }}` outside of the `<div>` tag makes it look this way - 
screenshot 2018-12-01 at 18 47 39 which is a bit better IMO. which one should I use?

Comment (atodorov) [reply tcms/testplans/templates/testplans/mutable.html]: None of these looks like the one in `testruns/mutable.html`. The red field is the same size as the input element. Somehow you have mismatch where you place the error markup.  Here's my diff and screenshot:

```
             <div class="form-group">
                 <label class="col-md-1 col-lg-1" for="id_name">{% trans "Name" %}</label>
-                <div class="col-md-11 col-lg-11">
+                <div class="col-md-11 col-lg-11 {% if form.name.errors %}has-error{% endif %}">
                     <input type="text" id="id_name" name="name" value="{% firstof form.name.value "" %}" class="form-control">
+                    {{ form.name.errors }}
                 </div>
             </div>
```

tp_name_error

File: tcms/testplans/static/testplans/js/mutable.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    });
+
+    $('#add_id_version').click(function () {
+        return showRelatedObjectPopup(this);
+    });
+}));
+
+function populateProductVersion() {
+    $('#id_version').find('option').remove();
+    $('#add_id_version').get(0).href += `&product=${$('#id_product').val()}`;
```
Comment (atodorov) [inline tcms/testplans/static/testplans/js/mutable.js]: This will not work as expected. Try selecting a different Product a few times without changing the Version. The new parameter will be appended to the URL, not replaced. You have to make sure there is only 1 `product=X` parameter to the URL.




Also `get(0)` I more often written as `$('selector')[0]` to get the first element.

Comment (asankov) [reply tcms/testplans/static/testplans/js/mutable.js]: Again, good catch. I fixed it with a dirty hack.

File: tcms/testplans/static/testplans/js/mutable.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+        return showRelatedObjectPopup(this);
+    });
+}));
+
+function populateProductVersion() {
+    let href = $('#add_id_version')[0].href;
+    $('#add_id_version')[0].href = href.slice(0, href.indexOf('&product'));
+
+    $('#id_version').find('option').remove();
+    $('#add_id_version')[0].href += `&product=${$('#id_product').val()}`;
```
Comment (atodorov) [inline tcms/testplans/static/testplans/js/mutable.js]: swap these 2 lines so that the operations with the href attribute are grouped together.

File: tcms/testplans/static/testplans/js/mutable.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    });
+
+    $('#add_id_version').click(function () {
+        return showRelatedObjectPopup(this);
+    });
+}));
+
+function populateProductVersion() {
+    let href = $('#add_id_version')[0].href;
+    $('#add_id_version')[0].href = href.slice(0, href.indexOf('&product'));
```
Comment (atodorov) [inline tcms/testplans/static/testplans/js/mutable.js]: this unfortunately doesn't produce the desired effect when the link doesn't contain the '&product' sub-string. As-is the index will be -1 and you end up removing the last character so the link becomes `?_popu&product=x`.

The easiest way to fix this is to add `&product=` to the URL in the template and you will not have to use conditional statements in this function.

File: tcms/testplans/templates/testplans/mutable.html [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+            <div class="form-group">
+                <div class="col-md-1 col-lg-1">
+                    <label for="id_product">{% trans "Product" %}</label>
+                    <a href="{% url 'admin:management_product_add' %}?_popup" id="add_id_product" alt="{% trans 'add new Product' %}" title="{% trans 'add new Product' %}">+</a>
+                </div>
+                <div class="col-md-3 col-lg-3 {% if form.product.errors %}has-error{% endif %}">
+                    <select id="id_product" name="product" class="form-control">
+                        {% for product in form.product.field.queryset %}
+                            <option value="{{ product.pk }} {% if product.pk == form.product.value %}selected{% endif %}">
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: The closing `"` must be before the if statement

File: tcms/testplans/templates/testplans/mutable.html [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+                    <label for="id_version">{% trans "Version" %}</label>
+                    <a href="{% url 'admin:management_version_add' %}?_popup&product={{ form.product.pk }}"
+                       id="add_id_version" alt="{% trans 'add new Version' %}" title="{% trans 'add new Version' %}">
+                        +
+                    </a>
+                </div>
+                <div class="col-md-3 col-lg-3 {% if form.product_version.errors %}has-error{% endif %}">
+                    <select id="id_version" name="product_version" class="form-control">
+                        {% for product_version in form.product_version.field.queryset %}
+                            <option value="{{ product_version.pk }} {% if product_version.pk == form.product_version.value %}selected{% endif %}">
```
Comment (atodorov) [inline tcms/testplans/templates/testplans/mutable.html]: The closing `"` must be before the if statement
